src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -2362,148 +2392,115 @@ private void InferTypeFrom(VariableExpressionAst variableExpressionAst, List<PST
+            {
+                ScopeIsLocal = true,
+                LocalScopeOnly = variableExpressionAst.VariablePath.IsLocal || variableExpressionAst.VariablePath.IsPrivate,
+                StopSearchOffset = variableExpressionAst.Extent.StartOffset,
+                VariableTarget = variableExpressionAst
+            };
+            while (parent is not null)
             {
-                if (ast is ParameterAst parameterAst)
+                if (parent is IParameterMetadataProvider)
                 {
-                    var currentCount = inferredTypes.Count;
-                    inferredTypes.AddRange(InferTypes(parameterAst));
+                    parent.Visit(assignmentVisitor);
 
-                    if (inferredTypes.Count != currentCount)
+                    if (assignmentVisitor.LastConstraint is not null)
                     {
-                        return;
+                        inferredTypes.Add(new PSTypeName(assignmentVisitor.LastConstraint));
+                        break;
                     }
-                }
-            }
 
-            var assignAsts = targetAsts.OfType<AssignmentStatementAst>().ToArray();
+                    if (assignmentVisitor.LocalScopeOnly)
+                    {
+                        break;
+                    }
 
-            // If any of the assignments lhs use a type constraint, then we use that.
-            // Otherwise, we use the rhs of the "nearest" assignment
-            for (int i = assignAsts.Length - 1; i >= 0; i--)
-            {
-                if (assignAsts[i].Left is ConvertExpressionAst lhsConvert)
-                {
-                    inferredTypes.Add(new PSTypeName(lhsConvert.Type.TypeName));
-                    return;
+                    assignmentVisitor.ScopeIsLocal = false;
+                    assignmentVisitor.StopSearchOffset = parent.Extent.StartOffset;
                 }
-            }
 
-            var foreachAst = targetAsts.OfType<ForEachStatementAst>().FirstOrDefault();
-            if (foreachAst != null)
-            {
-                inferredTypes.AddRange(
-                    GetInferredEnumeratedTypes(InferTypes(foreachAst.Condition)));
-                return;
-            }
-
-            var commandCompletionAst = targetAsts.OfType<CommandAst>().FirstOrDefault();
-            if (commandCompletionAst != null)
-            {
-                inferredTypes.AddRange(InferTypes(commandCompletionAst));
-                return;
+                parent = parent.Parent;
             }
 
-            int smallestDiff = int.MaxValue;
-            AssignmentStatementAst closestAssignment = null;
-            foreach (var assignAst in assignAsts)
+            if (assignmentVisitor.LastConstraint is null)
             {
-                var endOffset = assignAst.Extent.EndOffset;
-                if ((startOffset - endOffset) < smallestDiff)
+                if (assignmentVisitor.LastAssignment is not null)
```

Comment:
Maybe swap blocks in the `if` for readability:
```c#
                if (assignmentVisitor.LastAssignment is null)
                {
                    if (assignmentVisitor.LastAssignmentType is not null)
                    {
                        inferredTypes.Add(assignmentVisitor.LastAssignmentType);
                    }
                } else ...
```

Reply:
I think the current setup is easier to read with less nesting and a consistent "is not null" check for both scenarios.

Reply:
My idea is only to move short block on top for readability.

Reply:
I ended up removing the outer `if` as I had to make some other changes as well.